### PR TITLE
[spirv] Support output disassembly in dxc

### PR DESCRIPTION
**Pull request by antiagainst**

### CODE REVIEW

**Comment by antiagainst** [comment]
@ehsannas PTAL

**Comment by antiagainst** [comment]
This pull request touches source code for the dxc tool.

File: tools/clang/tools/dxc/dxc.cpp [hunk 1]
```diff
@@ -62,6 +62,43 @@
...
+
+  std::vector<uint32_t> words;
+  words.resize(num32BitWords, 0);
+  memcpy(words.data(), binaryStr.data(), binaryStr.size());
+
+  std::string assembly;
+  spvtools::SpirvTools spirvTools(SPV_ENV_VULKAN_1_0);
+  spirvTools.SetMessageConsumer(
+    [](spv_message_level_t, const char *, const spv_position_t &,
+      const char *message) { fprintf(stderr, "%s\n", message); });
```
Comment (marcelolr) [inline tools/clang/tools/dxc/dxc.cpp]: fprintf might not be what you want if the user has a console with a different encoding. Just something to think about - you may or may not care about this.

File: tools/clang/tools/dxc/dxc.cpp [hunk 2]
```diff
@@ -274,6 +311,25 @@ int DxcContext::ActOnBlob(IDxcBlob *pBlob, IDxcBlob *pDebugBlob, LPCWSTR pDebugB
...
+#ifdef ENABLE_SPIRV_CODEGEN
+  if (m_Opts.GenSPIRV) {
+    CComPtr<IDxcLibrary> pLibrary;
+    IFT(m_dxcSupport.CreateInstance(CLSID_DxcLibrary, &pLibrary));
+
+    if (DisassembleSpirv(pBlob, pLibrary, &pDisassembleResult)) {
+      if (!m_Opts.AssemblyCode.empty()) {
+        WriteBlobToFile(pDisassembleResult, m_Opts.AssemblyCode);
+      } else {
+        WriteBlobToConsole(pDisassembleResult);
```
Comment (marcelolr) [inline tools/clang/tools/dxc/dxc.cpp]: You can add a test to the /utils/hct/hcttestcmds.cmd file to have it succeed on errorlevel 0 or if we find the "not available" string in the output.

**Comment by marcelolr** [comment]
@antiagainst thanks for the diligence. I've added some comments; the "raw pointer" and test coverage are probably the more interesting ones; handling console encoding is a bit of a mess oftentimes (especially if you're concerned about console vs. file output) so perhaps you want to punt on that.

File: tools/clang/tools/dxc/dxc.cpp [hunk 3]
```diff
@@ -62,6 +62,43 @@
 #include <algorithm>
 #include <unordered_map>
 
+// SPIRV Change Starts
+#ifdef ENABLE_SPIRV_CODEGEN
+#include "spirv-tools/libspirv.hpp"
+
+static bool DisassembleSpirv(const CComPtr<IDxcBlob> &binaryBlob,
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply tools/clang/tools/dxc/dxc.cpp]: TIL. Thanks! I've changed to use raw pointers.

File: tools/clang/tools/dxc/dxc.cpp (hunk 1 above)
Comment (antiagainst) [reply tools/clang/tools/dxc/dxc.cpp]: Removed. We don't really care error messages here (unlikely to happen anyway).

File: tools/clang/tools/dxc/dxc.cpp (hunk 2 above)
Comment (antiagainst) [reply tools/clang/tools/dxc/dxc.cpp]: Done. Not sure what I've done is the typical batch script way. I tried to do the `findstr` in the first if check but it seems `errorlevel` will not be updated at all. Haven't figured out why.

**Comment by antiagainst** [comment]
@marcelolr: I'm not sure how your settings are. But `hcttestcmds.cmd` does not work on my machine since it cannot find `smoke.hlsl`. I've changed that to use absolute path. Let me know if that's not the suitable way.

**Comment by marcelolr** [comment]
Minor update to make env var local to script, I've modified in-place to avoid round-trip, will merge once AppVeyor gives me a thumbs-up. Thanks @antiagainst !

**Comment by antiagainst** [comment]
Thanks @marcelolr! BTW, I've revised my second commit multiple times and it seems Appveyor is not cancelling the job for the previous iterations. You may want to log into Appveyor to manually cancel all the previous iterations except the last one (1.0.890) for this pull request. :)

**Pull request merged**